pilot/pkg/xds/ads.go

```diff
@@ -601,32 +602,49 @@ func (s *DiscoveryServer) initProxyMetadata(node *core.Node) (*model.Proxy, erro
 // setTopologyLabels sets locality, cluster, network label
 // must be called after `SetWorkloadLabels` and `SetServiceInstances`.
 func setTopologyLabels(proxy *model.Proxy) {
-	var localityStr string
-	// Get the locality from the proxy's service instances.
-	// We expect all instances to have the same IP and therefore the same locality.
-	// So its enough to look at the first instance.
-	if len(proxy.ServiceInstances) > 0 {
-		localityStr = proxy.ServiceInstances[0].Endpoint.Locality.Label
-	} else {
-		// If no service instances(this maybe common for a pure client), respect LocalityLabel
-		localityStr = proxy.Labels[model.LocalityLabel]
+	// This is a bit un-intuitive, but pull the locality from Labels first. The service registries have the best access to
+	// locality information, as they can read from various sources (Node on Kubernetes, for example). They will take this
+	// information and add it to the labels (via AugmentLabels). So while the proxy may not originally have these labels,
+	// it will by the time we get here (as a result of calling this after SetWorkloadLabels). Note: this may also be from a
+	// previous iteration of setTopologyLabels on the proxy. This is fine; if the service registry has updated information
+	// it would override the labels, else we use the previously known value.
+	locality := localityFromProxyLabels(proxy)
+	if locality == nil {
+		// If no labels set, we didn't find the locality from the service registry. We do support a (mostly undocumented/internal)
+		// label to override the locality, so respect that here as well.
+		ls, f := proxy.Labels[model.LocalityLabel]
```

Comment:
This is also from labels, can we move it to `localityFromProxyLabels`

Reply:
Ah if you take a look at `getPodLocality`，you will find that the `istio-locality` label will take priority over node label: [NodeRegionLabelGA,NodeZoneLabelGA, and topology.istio.io/subzone]

Reply:
thanks, resolving as this is moved by https://github.com/istio/istio/pull/46326